araig_test_runners/scripts/test_4_static_obstacle

```diff
@@ -0,0 +1,136 @@
+    "test_succeeded"       : "/signal/runner/test_succeeded",
+    "start_test"           : "/signal/ui/start_test",
+    "reset_test"           : "/signal/ui/reset_test",
+    "interrupt_test"       : "/signal/ui/interrupt_test",
+}
+"""
+
+class State(enum.Enum):
+    wait = 1
+    start_test = 2
+    interrupt_test = 3
+    complete_test = 4
+    check_condition = 5
+    reached_goal = 6
+
+class Test4(TestBase):
+    def __init__(self, rate):
+        
+        extend_subscribers_dict = {
+            "goal"                  : "/signal/calc/robot_reached_goal_and_stopped",
+            "robot_in_collision"    : "/signal/calc/robot_in_collision"
+        }
+        extend_publishers_dict = {
+        }
+        extend_param_list = {
+            "timeout_check_goal",
+            "timeout_after_complete"
+        }
+
+        self.pre_state = None
+        self.current_state = State.wait
+
+        super(Test4, self).__init__(sub_dict = extend_subscribers_dict, \
+        pub_dict = extend_publishers_dict, \
+        param_list = extend_param_list, \
+        rate = rate)
+
+    def main(self):
+        if self.pre_state == None and self.current_state == State.wait:
+            rospy.loginfo(rospy.get_name() + ": Waiting to start...")
+            self.pre_state = State.wait
+
+         # start test, start robot (send goal)
+        if self.current_state == State.wait and self.getSafeFlag("start_test"):
+            self.startRecordingAndWait(3)
+            rospy.loginfo(rospy.get_name() + ": Starting robot.")
+            self._publishers["start_robot"].publish(self.buildNewBoolStamped(True))
+            self.current_state = State.start_test
+            self.pre_state = State.start_test
+
+        # robot in collision zoom
+        if self.current_state == State.start_test and self.getSafeFlag("robot_in_collision"):
+            rospy.loginfo(rospy.get_name() + ": Robot reached collision zoom , ending test in failure!")
+            self._publishers["test_failed"].publish(self.buildNewBoolStamped(True))
+            self._publishers["test_completed"].publish(self.buildNewBoolStamped(True))
+            self.current_state = State.complete_test
+
+        # If interrupted
+        if self.current_state == State.start_test and self.getSafeFlag("interrupt_test"):
+            rospy.logwarn(rospy.get_name() + ": Interrupted!!")
+            self._publishers["test_failed"].publish(self.buildNewBoolStamped(True))
+            self._publishers["test_completed"].publish(self.buildNewBoolStamped(True))
+            # !!! ui should publish interrupt_test to False after pub True, interrupt_test signal is event 
+            self._publishers["interrupt_test"].publish(self.buildNewBoolStamped(False))
+            self.current_state = State.complete_test
+
+        # If robot has stopped (because it found an obstacle), but maybe robot will still move
+        if self.current_state == State.start_test and self.getSafeFlag("goal"):
```

Comment:
okay checking for oscillations. I get why you do this, but as before, the test itself should not try to compensate for oscillations. So the goal time checking is done immediately as the robot reaches goal, as that is the specification of this test. Checking for oscillations is an addendum, and if you remember our old discussion, we can just count the number of times the robot has oscillated and show that in the UI/log it.

In general, tests should not be altered to accommodate for shortcomings in the test subjects.

Reply:
Checking for oscillations is good. So if does oscillate, then count the number, and report a third result category: test_oscillated.   
 
Verify/implement a timeout to number of oscillations that the runner can check. If more, then fail.   

Finally, log the number of oscillations. (Figure out how)